lambda-functions/elb-cleanup/aws.go

```diff
@@ -0,0 +1,116 @@
+package main
+
+import (
+	"context"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/elb"
+	"github.com/aws/aws-sdk-go/service/elbv2"
+	"github.com/pkg/errors"
+)
+
+// awsTimeout used for the context of AWS SDK
+const awsTimeout = 50 * time.Second
+
+// Client for making AWS requests
+type Client struct {
+	// ec2 *ec2.EC2
+	elbv2 *elbv2.ELBV2
+	elb   *elb.ELB
+}
+
+// Resourcer the interface for the AWS client
+type Resourcer interface {
+	ListUnusedElbs(context context.Context) ([]elbv2.LoadBalancer, error)
+	DeleteElbs(context context.Context, unUsedLBs []elbv2.LoadBalancer) error
+	ListUnUsedClassiclbs(context context.Context) ([]*elb.LoadBalancerDescription, error)
+	DeleteClassiclbs(context context.Context, elbDescList []*elb.LoadBalancerDescription) error
+}
+
+// NewClient factory method to create AWS client
+func NewClient(sess *session.Session) *Client {
+	return &Client{
+		elbv2: elbv2.New(sess),
+		elb:   elb.New(sess),
+	}
+}
+
+// ListUnusedElbs it will find any unused network LB
+func (c *Client) ListUnusedElbs(context context.Context) ([]elbv2.LoadBalancer, error) {
+	input := &elbv2.DescribeLoadBalancersInput{
+		LoadBalancerArns: []*string{},
+	}
+
+	result, err := c.elbv2.DescribeLoadBalancers(input)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed elbv2.DescribeLoadBalancer")
+	}
+
+	var unUsedLBs []elbv2.LoadBalancer
+	var isUnused bool
+	for _, lb := range result.LoadBalancers {
+		isUnused = true
+		targetGroups, err := c.elbv2.DescribeTargetGroups(&elbv2.DescribeTargetGroupsInput{LoadBalancerArn: lb.LoadBalancerArn})
+		if err != nil {
+			return nil, errors.Wrap(err, "failed elbv2.DescribeTargetGroups")
+		}
+
+		for _, targetGroup := range targetGroups.TargetGroups {
```

Comment:
Because I am curious, target groups can exist and can be orphans which means unused ELB? Or in practice LBs are totally empty without `TargetGroups`?

Reply:
Good Question, Yes currently criteria for unused ELB is that associated Target group exist with no registered targets irrespective of target instance health. e.g if an ELB exist even with all un-healthy instances then it won't qualify for unused ELB.
We can easily extend it to cover use case of ELB without any target group but that is very rare case to exist.